File: app/src/main/java/me/kirimin/mitsumine/model/Bookmark.java
package me.kirimin.mitsumine.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

public class Bookmark implements Parcelable {

    private final String user;
    private final List<String> tags;
    private final String timeStamp;
    private final CharSequence comment;
    private final String userIcon;
    private boolean isPrivate;

    public Bookmark(String user, List<String> tags, String timeStamp, CharSequence comment, String userIcon) {
        this.user = user;
        this.tags = tags;
        this.timeStamp = timeStamp;
        this.comment = comment;
        this.userIcon = userIcon;
    }

    public String getUser() {
        return user;
    }

    public List<String> getTags() {
        return tags;
    }

    public String getTimeStamp() {
        return timeStamp;
    }

    public CharSequence getComment() {
        return comment;
    }

    public String getUserIcon() {
        return userIcon;
    }

    public boolean isPrivate() {
        return isPrivate;
    }

    public void setPrivate(boolean isPrivate) {
        this.isPrivate = isPrivate;
    }

    protected Bookmark(Parcel in) {
        user = in.readString();
        if (in.readByte() == 0x01) {
            tags = new ArrayList<>();
            in.readList(tags, String.class.getClassLoader());
        } else {
            tags = null;
        }
        timeStamp = in.readString();
        comment = (CharSequence) in.readValue(CharSequence.class.getClassLoader());
        userIcon = in.readString();
        isPrivate = in.readByte() != 0x00;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(user);
        if (tags == null) {
            dest.writeByte((byte) (0x00));
        } else {
            dest.writeByte((byte) (0x01));
            dest.writeList(tags);
        }
        dest.writeString(timeStamp);
        dest.writeValue(comment);
        dest.writeString(userIcon);
        dest.writeByte((byte) (isPrivate ? 0x01 : 0x00));
    }

    @SuppressWarnings("unused")
    public static final Parcelable.Creator<Bookmark> CREATOR = new Parcelable.Creator<Bookmark>() {
        @Override
        public Bookmark createFromParcel(Parcel in) {
            return new Bookmark(in);
        }

        @Override
        public Bookmark[] newArray(int size) {
            return new Bookmark[size];
        }
    };
}

File: app/src/main/java/me/kirimin/mitsumine/domain/model/Bookmark.kt
package me.kirimin.mitsumine.domain.model

import android.os.Parcel
import android.os.Parcelable

import java.util.ArrayList

public class Bookmark : Parcelable {

    public val user: String
    public val tags: List<String>
    public val timeStamp: String
    public val comment: CharSequence
    public val userIcon: String
    private var isPrivate: Boolean = false

    public constructor(user: String, tags: List<String>, timeStamp: String, comment: CharSequence, userIcon: String) {
        this.user = user
        this.tags = tags
        this.timeStamp = timeStamp
        this.comment = comment
        this.userIcon = userIcon
    }

    public fun isPrivate(): Boolean {
        return isPrivate
    }

    public fun setPrivate(isPrivate: Boolean) {
        this.isPrivate = isPrivate
    }

    public fun hasComment(): Boolean {
        return !comment.toString().isEmpty()
    }

    protected constructor(`in`: Parcel) {
        user = `in`.readString()
        if (`in`.readByte() == 1.toByte()) {
            tags = ArrayList<String>()
            `in`.readList(tags, javaClass<String>().getClassLoader())
        } else {
            tags = ArrayList()
        }
        timeStamp = `in`.readString()
        comment = `in`.readValue(javaClass<CharSequence>().getClassLoader()) as CharSequence
        userIcon = `in`.readString()
        isPrivate = `in`.readByte() != 0.toByte()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(user)
        dest.writeByte((1).toByte())
        dest.writeList(tags)
        dest.writeString(timeStamp)
        dest.writeValue(comment)
        dest.writeString(userIcon)
        dest.writeByte((if (isPrivate) 1 else 0).toByte())
    }

    companion object {
        public val CREATOR: Parcelable.Creator<Bookmark> = object : Parcelable.Creator<Bookmark> {
            override fun createFromParcel(`in`: Parcel): Bookmark {
                return Bookmark(`in`)
            }

            override fun newArray(size: Int): Array<Bookmark?> {
                return arrayOfNulls(size)
            }
        }
    }
}